test/gen_dune_rules.ml

```diff
@@ -0,0 +1,30 @@
+let is_testcase name =
+  Fpath.(has_ext "dot" (v name))
+
+let pp_targets ppf filenames =
+  Fmt.pf ppf
+    {|(rule
+ (targets %a)
+ (action (run ./test.exe)))
```

Comment:
that part looks a bit weird to me and looks like the core of the problem: `test.ml` is an alcotest test suite, but also outputs dot files that you diff afterwards. Is that correct?

If that's the case, I think it'll be easier to remove this hybrid approach either by:
- using an alcotest assert in `test.ml`. If you want a promote-based workflow, I suppose that using ppx_expect inside `test.ml` would work as well.
- or splitting completely the part that emits dot files from the alcotest test suite (not sure if possible)

Reply:
Yes, that's correct. There is an Alcotest suite that outputs dotfiles with some expected equivalent, and currently we're using Dune to diff them in order to benefit from a promote workflow.

I don't think it's possible to split the dot file emission from the Alcotest suite. I wasn't aware that it was possible to use `ppx_expect` to benefit from promotion; how would that work?

Reply:
The idea is that you write bits like that for every dot file:

```ocaml
let dot = generate_dot "test name" in
print_endline (Dot.to_string dot);
[%expect]
```

And it will diff your ml file against a generated ml file where the output of Dot.to_string has been spliced into the `[%expect]` node. It can then be promoted. (the dot file will be inside a string literal). It's a bit unusual to mix ppx_expect and alcotest like that but should work.

Reply:
I see; nice. In this case, I think the current solution is probably good enough without taking on an additional test dependency. My only real issue with it is solved by the `test-force` Makefile target. 

Having the expected outputs separate from the test code is nice because it makes it easy to inspect them by e.g. converting a particular file to SVG.

Reply:
after giving it a bit more thought, combining alcotest and ppx_expect wouldn't work as expectations need to run under a toplevel `let%expect_test ...`.

It seems that what you're after is `Alcotest.check` where the `expected` part is a string literal that's backed by an external file which knows about promotion. Which is not really possible today unfortunately.